src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs

```diff
@@ -132,99 +137,128 @@ public virtual byte[] GetValue()
 
         internal static void TransformControls(DirectoryControl[] controls)
         {
+            Span<byte> attributeNameScratchSpace = stackalloc byte[AttributeNameStackAllocationThreshold];
+
             for (int i = 0; i < controls.Length; i++)
             {
                 Debug.Assert(controls[i] != null);
-                byte[] value = controls[i].GetValue();
+                byte[] value = controls[i]._directoryControlValue ?? Array.Empty<byte>();
+                Span<byte> asnSpan = value;
+                bool asnReadSuccessful;
+
                 if (controls[i].Type == "1.2.840.113556.1.4.319")
                 {
-                    // The control is a PageControl.
-                    object[] result = BerConverter.Decode("{iO}", value);
-                    Debug.Assert((result != null) && (result.Length == 2));
+                    // The control is a PageControl, as described in RFC 2696.
+                    byte[] cookie = Array.Empty<byte>();
+
+                    AsnDecoder.ReadSequence(asnSpan, AsnEncodingRules.BER, out int sequenceContentOffset, out int sequenceContentLength, out _);
+                    Debug.Assert(sequenceContentLength > 0);
 
-                    int size = (int)result[0];
-                    // user expects cookie with length 0 as paged search is done.
-                    byte[] cookie = (byte[])result[1] ?? Array.Empty<byte>();
+                    asnReadSuccessful = AsnDecoder.TryReadInt32(asnSpan.Slice(sequenceContentOffset), AsnEncodingRules.BER, out int size, out int bytesConsumed);
```

Comment:
AsnDecoder.ReadSequence doesn't check that the contents don't "reach out" (particularly if a definite length encoding is used).  So, if for some reason it's not viable to check that the sequence was the full byte range earlier, this needs to use a sliced down input.  `asnSpan.Slice(sequenceContentOffset, sequenceContentLength)` being the most obvious, but that makes things harder for the next call; so really you'd just do

```suggestion
                    asnSpan = asnSpan.Slice(sequenceContentOffset, sequenceContentLength);
                    asnReadSuccessful = AsnDecoder.TryReadInt32(asnSpan, AsnEncodingRules.BER, out int size, out int bytesConsumed);
```